### ISSUE TITLE
How to make progressbar width same as output window in console

### ISSUE THREAD

**Issue by notuxius** [tqdm/tqdm#41]
or even adjustable on resizing

**Comment by lrq3000** [comment]
You can specify the `ncols` argument to an integer of your choice. Try different values and see for yourself which one allows the bar to fit your output window.

Not sure we can do anything about window resizing, I'm not sure we can detect that from Python (at least without using hacky per console tricks).

**Comment by casperdcl** [comment]
My implementation attempts to automatically detect terminal width on a number of different systems. However I only do this once upon initialisation to avoid the overhead of checking for resizes. If enough people request this I may implement a `dynamic_resize` bool argument.

**Comment by BertrandBordage** [comment]
I don’t think anyone would find that overhead too heavy. In all the cases I use tqdm, I rarely resize a window, so I don’t really need this feature, but if it takes 1s more per hour of execution, it’s ok, I vote for adding it and having it enabled by default. Of course, being able to disable it could be nice, especially for tiny machines like raspberries.

Anyway, that’s very nice of you to try to have everything optimized, I’m the same ^^

**Comment by lrq3000** [comment]
@BertrandBordage This kind of dynamic resizing will probably take up a lot more than just 1s more per hour of execution, because this requires not only additional computations **at each iteration**, but also it has to call IO libraries, and any IO operation is a bottleneck (IO is the slowest kind of operation a program can do).
Maybe I'm wrong, but if this is correct, then we could still add the feature but not enable it by default.

**Comment by BertrandBordage** [comment]
@lrq3000 Using [this solution](http://stackoverflow.com/a/566752/1576438), it takes 5 μs to get terminal size. Given that there’s by default at max an iteration every 0.1 s, this gives a maximum slowdown of 50 μs / s, equivalent to 0.18 s / h.

Sorry, but that’s far from slowing down 10 times the program ;)

**Comment by lrq3000** [comment]
@BertrandBordage : [tqdm's overhead is about 60ns/it](https://github.com/tqdm/tqdm/blob/master/README.rst) (depends on the machine specs of course), so a slowdown of 50 microsec/it equals 50000 ns/it, so it would be a slowdown of about 1000x. In reality, I think it will slowdown a bit less but still it will probably be a big slowdown.

**Comment by lrq3000** [comment]
I can be wrong however, but I doubt that considering it requires a IO request. The only way the IO request could be speedy enough as to not slowdown tqdm would be if the terminal caches its size with some mechanism, but I doubt such a cache mechanism exists, and even if it exists, it certainly does not exist on all platforms (I bet on Linux, but I certainly won't on Windows' shell).

But anyway as I said, we can still implement this feature, it's an interesting one, but I don't think we can set it as default because of the speed penalty (this can be confirmed using the tqdm benchmarks).

**Comment by casperdcl** [comment]
Meh. The slow down won't be significant for loops with few (heavyweight) iterations. I'll add support and then we can all go wild altering `_environ_cols` to play nice on all the different combinations of python and OS flavours. :D

**Comment by casperdcl** [comment]
But definitely don't think enabled by default is a good idea

**Comment by BertrandBordage** [comment]
Ah, I see why you don’t see my point: the terminal size only need to be measured on each tqdm display, not on each tqdm iteration. It doesn’t impact all tqdm iterations, except if your script iterations are slower than 0.1s. The cost of measuring the terminal size is nothing compared to the cost of actually writing to stderr, as you can see here:
```python
In [1]: import sys
In [2]: %timeit sys.stderr.write('something\r')
100000 loops, best of 3: 24.3 µs per loop
```

To get a good estimation of the real cost of this issue we need to use an optimized version of that function, like this:
```python
from fcntl import ioctl
from termios import TIOCGWINSZ
from struct import pack, unpack

FOUR_UINT16_ZEROS = pack('HHHH', 0, 0, 0, 0)

def get_terminal_size():
    h, w, hp, wp = unpack('HHHH', ioctl(0, TIOCGWINSZ, FOUR_UINT8_ZEROS))
    return w, h
```

It takes 1 μs / iteration instead of 5 μs / iteration compared to the previous solution I mentioned.
So it will only slow down the user script of a maximum of `(1e-6 / 0.1) * 3600` → 36 ms / h, which is perfectly acceptable.

Besides, I don’t know how a 60 ns/it cost was measured, but I measure **a 6 to 3500 × higher cost**:

First use case:
- 1000 iterations
- one display
- **cost of tqdm: 363 ns / it**
- estimated cost if there was dynamic resizing: `((363e-9 * 1000) + (1e-6 * 1)) / 1000` ⇒ 364 ns / it

Second use case:
- 10 iterations
- 10 displays
- **cost of tqdm: 211 μs / it**
- estimated cost if there was dynamic resizing: `((211e-6 * 10) + (1e-6 * 10)) / 10` ⇒ 212 μs / it

**So having dynamic resizing enabled by default has a 0.27 to 0.48 % impact on tqdm performance, which is ridiculously acceptable.**

Here’s the detail:
```python
In [1]: from tqdm import tqdm
In [2]: %%timeit -o
        for i in tqdm(range(1000)):
            pass
1000 loops, best of 3: 385 µs per loop                                          

In [3]: %%timeit -o
        for i in range(1000):
            pass
10000 loops, best of 3: 21.3 µs per loop

In [4]: (__.best - _.best) / 1000
Out[4]: 3.6360087394714355e-07
```
```python
In [1]: from tqdm import tqdm
In [2]: from time import sleep
In [3]: %%timeit -o
        for i in tqdm(range(10)):
            sleep(0.1)
1 loops, best of 3: 1 s per loop                                                

In [4]: %%timeit -o
        for i in range(10):
            sleep(0.1)
1 loops, best of 3: 1 s per loop

In [5]: (__.best - _.best) / 10
Out[5]: 0.00021119117736816405
```

**Comment by casperdcl** [comment]
You're (mostly) measuring startup overhead, not per-iteration overhead. The per-iteration overhead is what is quoted in the README and this is what we are focused on keeping low.

**Comment by BertrandBordage** [comment]
I don’t think so, if I remove the startup time on a long running script like the second case in my previous message, I get similar results:
```python
In [1]: from tqdm import tqdm
In [2]: from time import sleep
In [3]: %timeit -o tqdm(range(1000))
  0%|                                                  | 0/1000 [00:00<?, ?it/s]10000 loops, best of 3: 202 µs per loop

In [4]: %%timeit -o
        for i in tqdm(range(10)):
            sleep(0.1)
1 loops, best of 3: 1 s per loop

In [5]: %%timeit -o
        for i in range(10):
            sleep(0.1)
1 loops, best of 3: 1 s per loop

In [6]: (__.best - ___.best - _.best) / 10
Out[6]: 0.0002824415206909237
```
We get **282 μs**, which is on this run even worse than the previous results. And the real cost is higher since removing the startup time also removes the cost of the first display (which by the way should not happen on initialization I guess).

Of course, if I remove the startup time on the first case of my previous message, it’s way faster because we removed the timing of the only display happening during the whole loop.

**Comment by BertrandBordage** [comment]
If the measure of 60 ns / it was done using [simple_examples.py](https://github.com/tqdm/tqdm/blob/367723c584d7bbbce0b24f079f9b0404dfb1758e/examples/simple_examples.py), then you’re measuring about 100 displays for 1e8 iterations. In the real world, there will be logic in the loop, so you’ll have a way lower number of iterations for the same amount of display, like 100 displays for 100000 iterations or 100 displays for 100 iterations. Even if you put something extremely simple like `int('1') + int('1')` in the loop, you get 10 times less iterations per second.

To sum up, you’re not measuring the cost of tqdm, you’re measuring the minimum cost it can have, when it is useless. In a normal case, it has up to about 3500 × that cost (at the default refresh rate of 10 displays / s).

**Comment by lrq3000** [comment]
@BertrandBordage : the "minimum cost it can have when it is useless" is exactly the overhead of tqdm: if you add inside your loop, of course your loop will take longer, but then it's not due to tqdm's own processing but to your instructions. Adding commands inside the loop won't change the overhead of tqdm since tqdm doesn't care about the content of your loop, it only cares about the iterable.

Anyway, let's just wait for @casperdcl to add the feature, and then we will just use the tqdm benchmarks to see if in practice the overhead is low enough to make it default.

**Comment by lrq3000** [comment]
The goal of tqdm is to have a negligible overhead compared to a loop without tqdm (ie: that the processing time inside your loop, whatever the commands are, will be of a computational complexity that is at least an order of magnitude higher than tqdm's complexity).

So, if the new feature has a negligible cost, maybe we will make it default, but we would prefer to keep the default options such as the cost is the most minimal possible, while still providing a good user experience out of the box. As long as the option can be enabled by just switching a boolean argument, I think any dev can use that easily to tailor tqdm to their needs. Tqdm cannot provide a perfectly fitted experience for everyone out-of-the-box, but it can provide options to allow that on a case by case basis.

**Comment by BertrandBordage** [comment]
@lrq3000: Sorry but no, it’s not the overhead. As I explained above, tqdm has two types of iterations: those where it counts, and those where it counts *and* displays.

You can’t say "tqdm takes 60ns per iteration". You need to say "tqdm takes 60ns per iteration when it displays nothing, and 211 μs when it displays something".

On normal cases, the ratio `displays count / iterations count` is close to 1, so you have an important tqdm overhead. On your naive benchmark, the ratio is 10 / 1e8, so you measure a way lower overhead.

**Comment by BertrandBordage** [comment]
You’ll see when it’ll be implemented, there will be no change :p

**Comment by lrq3000** [comment]
I understand now what you mean and I agree. We should mention that in the
readme (and in addition, this provides the opportunity to explain that the
overhead can be tweaked by specifying miniters and mininterval arguments).
Yes we'll see about the overhead of the new feature :)

2015-10-17 21:51 GMT+02:00 Bertrand Bordage <notifications@github.com>:

**Comment by casperdcl** [comment]
You're both right. `tqdm` will have an *effective* per-iteration overhead measurable in ns for inexpensive loops ranging up to something measurable in us for more demanding loops. In the latter case it is actually more likely that the end user won't care about the `tqdm` overhead as it would be fractionally small.

This is deliberately how `tqdm` was designed - to adapt in such a way that *fractional* overhead stays low for a range of scenarios.

Compromise: `dynamic_ncols` as a [default: False] option. As I said I suspect this will trigger further interest into optimising `_environ_cols` (I had implemented it as a quick hack and it has since remained untouched).

**Comment by BertrandBordage** [comment]
Sorry for all the long messages, I may have not explained it clearly from the beginning ^^

**Comment by lrq3000** [comment]
No problem, that's the internet! :p

Ok thank's Casper, we'll optimize it later if necessary :)

2015-10-17 22:29 GMT+02:00 Bertrand Bordage <notifications@github.com>:

**Comment by lrq3000** [comment]
Thank's Casper, seems all good but I cannot test on my computer because it's running Windows XD @BertrandBordage can you please test the new feature and see if it does what you wanted?

**Comment by casperdcl** [comment]
it works on my unix system. have you tried it on windows? ideally you can look into editing `_environ_cols` so it works on your computer, too

**Comment by lrq3000** [comment]
I'm not even sure if it's possible at all to detect the console size on Windows since it's a fake terminal (not running DOS anymore since Windows Vista) and with very limited functionalities. I'll take a look anyway.

**Comment by lrq3000** [comment]
Found a code snippet that builds on the link provided by @BertrandBordage with enhancements to support all platforms:
https://gist.github.com/jtriley/1108174

We could maybe implement that. As I thought, for Windows, we need to use ctypes, so we need a specific function. We can use `get_terminal_size()` function of the code snippet but instead of returning the terminal size, we could return a function pointer, so that the function would compute once which platform and which function to use. This way, we could have about the same overhead as now, but supporting all platforms.

**Comment by casperdcl** [comment]
Yes, my thoughts exactly

**Comment by casperdcl** [comment]
Since I have a dual boot I can test this.... in a couple of days.

**Comment by wheybags** [comment]
I realise this is an old issue, but couldn't you use a signal handler for SIGWINCH to detect this when necessary? This would avoid any overhead at all when not actually resizing the window.
Ofc, this would be a unix-only thing, but windows might have something similar.
